mc-sema/runtime/linux_i386_callback.c

```diff
@@ -0,0 +1,728 @@
+
+    // do memcpy
+    __asm__ volatile("cld\n");
+    __asm__ volatile("rep; movsb\n");
+
+    // call translated_function(reg_state);
+    __asm__ volatile("pushl %0\n": : "m"(__mcsema_do_call_state) );
+    __asm__ volatile("call *%eax\n");
+
+    // restore registers
+    __asm__ volatile("movl %0, %%ebx\n": : "m"(__mcsema_do_call_state.EBX) );
+    __asm__ volatile("movl %0, %%ecx\n": : "m"(__mcsema_do_call_state.ECX) );
+    __asm__ volatile("movl %0, %%edx\n": : "m"(__mcsema_do_call_state.EDX) );
+    __asm__ volatile("movl %0, %%esi\n": : "m"(__mcsema_do_call_state.ESI) );
+    __asm__ volatile("movl %0, %%edi\n": : "m"(__mcsema_do_call_state.EDI) );
+    __asm__ volatile("movl %0, %%ebp\n": : "m"(__mcsema_do_call_state.EBP) );
+    // *do not* restore ESP, although this may be a bug
+
+    // restore XMM
+    __asm__ volatile("movups %0, %%xmm0\n": : "m"(__mcsema_do_call_state.XMM0) );
+    __asm__ volatile("movups %0, %%xmm1\n": : "m"(__mcsema_do_call_state.XMM1) );
+    __asm__ volatile("movups %0, %%xmm2\n": : "m"(__mcsema_do_call_state.XMM2) );
+    __asm__ volatile("movups %0, %%xmm3\n": : "m"(__mcsema_do_call_state.XMM3) );
+    __asm__ volatile("movups %0, %%xmm4\n": : "m"(__mcsema_do_call_state.XMM4) );
+    __asm__ volatile("movups %0, %%xmm5\n": : "m"(__mcsema_do_call_state.XMM5) );
+    __asm__ volatile("movups %0, %%xmm6\n": : "m"(__mcsema_do_call_state.XMM6) );
+    __asm__ volatile("movups %0, %%xmm7\n": : "m"(__mcsema_do_call_state.XMM7) );
+
+    // save return value into rax
+    __asm__ volatile("movl %0, %%eax\n": : "m"(__mcsema_do_call_state.EAX) );
+
+    __asm__ volatile("retl\n");
+}
+
+typedef struct _do_call_state_t {
+    uint32_t __mcsema_real_esp;
+    uint32_t __mcsema_jmp_count;
+    char sse_state[512] __attribute__((aligned (16)));
+    uint32_t reg_state[15];
+} do_call_state_t;
+
+__thread do_call_state_t do_call_state[NUM_DO_CALL_FRAMES];
+__thread int32_t cur_do_call_frame = -1; /* XXX */
+__thread uint32_t call_frame_counter = 0; /* XXX */
+
+void do_call_value(void *state, uint32_t value)
+{
+    // get a clean frame to store state
+    ++cur_do_call_frame;
+    do_call_state_t *cs = &(do_call_state[cur_do_call_frame]);
+    call_frame_counter = 0;
+    cs->__mcsema_jmp_count = NUM_DO_CALL_FRAMES - cur_do_call_frame - 1;
+    //uint32_t reg_state[] = cs->reg_state;
+
+    __asm__ volatile(
+            "pusha\n" // save all regs just so we don't have bother keeping track of what we saved
+            "fxsave %0\n" // save sse state
+            "movl %3, %%eax\n"  // capture "state" arg (mcsema regstate)
+            "movl %4, %%ecx\n"  // capture "value" arg (call destination)
+            "movl %2, %%esi\n" // pointer to TLS area where we save state
+            "movl %c[state_edi](%%eax), %%edi\n" // dump struct regs to state
+            "movl %c[state_edx](%%eax), %%edx\n"
+            "movl %c[state_ebx](%%eax), %%ebx\n"
+            "movl %c[state_ebp](%%eax), %%ebp\n"
```

Comment:
suggest comment: ``// all further uses of asm variables must be constants` to makw the fact that using more generic asm operands could be unsafe as they might use `ebp` as a frame pointer. ALTERNATIVELY, add `ebp` to the clobber list.